File: main/scala/photos/brooklyn/practice/codingexercises/scala/leetcode/LongestMountain.scala
package photos.brooklyn.practice.codingexercises.scala.leetcode

/**
  * https://leetcode.com/contest/weekly-contest-87/problems/longest-mountain-in-array/
  */
object LongestMountain {
  def longestMountain(A: Array[Int]): Int = walkMountain(A)

  def walkMountain(arr: Array[Int], ascending: Int=0, descending: Int=0, index: Int=0, res: Int=0): Int =
    if(arr.length==index) res
    else (ascending, descending) match {
      // we are at the beginning
      case (0, _) => walkMountain(arr, 2, 0, index + 1)
      case (1, 1) => arr(index) match {
        case x if (x > ascending) => walkMountain(arr, 3, 0, index + 1)
        case x if (x == ascending) => walkMountain(arr, 4, 0, index + 1)
        case x if (x < ascending) => walkMountain(arr, 4, 1, index + 1)
      }
    }
}

File: main/java/photos/brooklyn/practice/codingexercises/java/leetcode/solutions/LongestMountain.java
package photos.brooklyn.practice.codingexercises.java.leetcode.solutions;

/**
 * https://leetcode.com/problems/longest-mountain-in-array/discuss/135593/C++JavaPython-1-pass-and-O(1)-space
 */
public class LongestMountain {
    public int longestMountain(int[] A) {
        // the peaks are where both are above zero, and valleys are where both are zero
        int res = 0, up = 0, down = 0;
        for (int i = 1; i < A.length; ++i) {
            if (down > 0 && A[i - 1] < A[i] || A[i - 1] == A[i]) up = down = 0;
            if (A[i - 1] < A[i]) up++;
            if (A[i - 1] > A[i]) down++;
            if (up > 0 && down > 0 && up + down + 1 > res) res = up + down + 1;
        }
        return res;
    }
}
